### DISCUSSION TITLE
Aggregating results across model points

### DISCUSSION THREAD

**Discussion by fumitoh** [lifelib-dev/lifelib#45] [Q&A]
Here's a response to comment on [Getting Started with lifelib - Part 5 - YouTube](https://www.youtube.com/watch?v=vqLcyZvKOo8&lc=UgzUJfOnAJDRuwE35bR4AaABAg):
"Is there a built-in function to aggregate across model points or is it up to the user to create a loop?"

BasicTerm_S projects cashflows for 1 model point at  a time. By default, Model point 1 is selected and its results are calculated.
To change the selected model point, assign the new model point ID to `point_id`.

```python

>>> import modelx as mx

>>> proj = mx.read_model(r"C:\path\to\BasicTerm_S").Projection    # need r before the path to treat backslashes literally

>>> proj.point_id         # Currently selected model point
1

>>> proj.pv_net_cf()
910.92066093366

>>> proj.point_id = 2   # Set the current model point to 2.  The previous results are cleared.

>>> proj.pv_net_cf()
1181.5470031411787

```
Here's how to loop through multiple model points, and aggregate the results of the model points.

```python

>>> sum_pvnetcf = 0

>>> for i in range(1, 11):
...     proj.point_id = i
...     sum_pvnetcf += proj.pv_net_cf()

>>> sum_pvnetcf
8745.014044652558
```

However, there's a much simpler approach. `Projection` is parameterized with `point_id`, so you can create a copy of Projection by  subscripting it.

```python
>>> proj.parameters
('point_id',)

>>> proj[1].pv_net_cf()     # proj(1).pv_net_cf() works as well
910.92066093366

>>> proj[2].pv_net_cf()
1181.5470031411787
```

So to get the same sum , you can just do:

```python
>>> sum(proj[i].pv_net_cf() for i in range(1,11))
8745.014044652558
```
Note that this approach is memory hungry as all the copies their values are kept.

**Comment by dogmom32** [comment]
Do you have any recommendations for speeding up the calculations in general? I basically modified the BasicTerm_S model to fit a specific product (e.g. add a more variables/calculations). This is a really cool library and very easy to code and reminds a lot of developing models in MG-ALFA. However, the speed is a concern and maybe it's just a python issue. 

I have a model that is a paid-up whole life with no premiums nor dividends. It takes about 12 hours to run through 37,000 policies/model points. Do you have any recommendations or would you like to see the model?

**Reply by fumitoh** [reply]
I want to take a look at the model and see if there's a quick fix. There're a couple of ways to speed up the calculations in general. I started [a new discussion](https://github.com/lifelib-dev/lifelib/discussions/46#discussion-4970775) since this is a new topic.

**Reply by fumitoh** [reply]
@jjaredrides Now modelx v0.23.0 can make BasicTerm_S 16 times faster. See https://modelx.io/blog/2023/08/19/enhanced-speed-for-exported-lifelib-models/
